### PR TITLE
Generating API reference docs for federation apiserver

### LINKED ISSUE #30541: Generate API reference docs for federation API
Similar to: http://kubernetes.io/docs/api-reference/v1/definitions/

Without it there is no way for users to figure out the structure for a Cluster object without having to look at our code.

cc @kubernetes/sig-cluster-federation

### PR DESCRIPTION
**Pull request by nikhiljindal**
Fixes #30541 

Adding a script `update-federation-api-reference-docs.sh` similar to the existing `update-api-reference-docs.sh` for kube-apiserver. Have moved the common parts to `hack/lib/swagger.sh`.
The new script will produce API reference docs for federation-apiserver.

Next step will be to surface these docs at kubernetes.io.

cc @kubernetes/sig-cluster-federation @kubernetes/sig-api-machinery @caesarxuchao

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31759)

### CODE REVIEW

**Comment by nikhiljindal** [comment]
@k8s-bot ok to test

**Comment by nikhiljindal** [comment]
URLs to view the generated files:

https://htmlpreview.github.io/?https://raw.githubusercontent.com/nikhiljindal/kubernetes/fedSwaggerDoc/federation/docs/api-reference/federation/v1beta1/operations.html

https://htmlpreview.github.io/?https://raw.githubusercontent.com/nikhiljindal/kubernetes/fedSwaggerDoc/federation/docs/api-reference/federation/v1beta1/definitions.html

https://htmlpreview.github.io/?https://raw.githubusercontent.com/nikhiljindal/kubernetes/fedSwaggerDoc/federation/docs/api-reference/extensions/v1beta1/operations.html

https://htmlpreview.github.io/?https://raw.githubusercontent.com/nikhiljindal/kubernetes/fedSwaggerDoc/federation/docs/api-reference/extensions/v1beta1/definitions.html

https://htmlpreview.github.io/?https://raw.githubusercontent.com/nikhiljindal/kubernetes/fedSwaggerDoc/federation/docs/api-reference/v1/operations.html

https://htmlpreview.github.io/?https://raw.githubusercontent.com/nikhiljindal/kubernetes/fedSwaggerDoc/federation/docs/api-reference/v1/definitions.html

File: federation/apis/swagger-spec/api.json [hunk 1]
```diff
@@ -3,6 +3,10 @@
   "apiVersion": "",
   "basePath": "https://10.10.10.10:6443",
   "resourcePath": "/api",
+  "info": {
```
Comment (caesarxuchao) [inline federation/apis/swagger-spec/api.json]: What is this?

File: hack/update-federation-generated-swagger-docs.sh [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+set -o pipefail
+
+KUBE_ROOT=$(dirname "${BASH_SOURCE}")/..
+source "${KUBE_ROOT}/hack/lib/init.sh"
+source "${KUBE_ROOT}/hack/lib/swagger.sh"
+
+kube::golang::setup_env
+
+GROUP_VERSIONS=(federation/v1beta1)
+GV_DIRS=("federation/apis/federation/v1beta1")
```
Comment (caesarxuchao) [inline hack/update-federation-generated-swagger-docs.sh]: maybe change to 
```bash
GV_DIRS="federation/$(kube::util::group-version-to-pkg-path "${group_version}")"
```

It's more consistent to the original script.

**Comment by caesarxuchao** [comment]
Review status: 0 of 22 files reviewed at latest revision, 5 unresolved discussions.

---

*[hack/update-api-reference-docs.sh, line 38 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31759#-KQTjy4AFu4MtJbwO5cL:-KQTjy4AFu4MtJbwO5cM:b-jhejdj) ([raw file](https://github.com/kubernetes/kubernetes/blob/cda78f8338cd1445d73dd74c862690fbd779f80f/hack/update-api-reference-docs.sh#L38)):*
> ```Shell
> 
> GROUP_VERSIONS=("v1" "extensions/v1beta1" "batch/v1" "autoscaling/v1" "certificates/v1alpha1")
> GV_DIRS=()
> ```

This looks to be copied from the original code, but I think using kube::util::group-version-to-pkg-path will make it cleaner.

---

*[hack/update-federation-api-reference-docs.sh, line 40 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31759#-KQTk7hK5JlcdC9ESAOj:-KQTk7hLviMEH2ThIKNn:bvxm2zs) ([raw file](https://github.com/kubernetes/kubernetes/blob/cda78f8338cd1445d73dd74c862690fbd779f80f/hack/update-federation-api-reference-docs.sh#L40)):*
> ```Shell
> GROUP_VERSIONS=("federation/v1beta1" "v1" "extensions/v1beta1")
> GV_DIRS=()
> for gv in "${GROUP_VERSIONS[@]}"; do
> ```

Same here, using kube::util::group-version-to-pkg-path seems better.

---

*[hack/lib/swagger.sh, line 79 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31759#-KQTke3HRTAmIwQpMBtD:-KQTke3HRTAmIwQpMBtE:b6ej391) ([raw file](https://github.com/kubernetes/kubernetes/blob/cda78f8338cd1445d73dd74c862690fbd779f80f/hack/lib/swagger.sh#L79)):*
> ```Shell
>   echo "Generating API reference docs for group versions: ${GROUP_VERSIONS[@]}, at dirs: ${GV_DIRS[@]}"
>   GROUP_VERSIONS=(${GROUP_VERSIONS[@]})
>   GV_DIRS=(${GV_DIRS[@]})
> ```

Are these two lines redundant?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31759)*

**Comment by caesarxuchao** [comment]
Just some minor comments and a question on what the generated code is.

---

Review status: 0 of 22 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31759#-:-KQTl3ISDKl5fNcmthJQ:b-b5ndy)*

**Comment by madhusudancs** [comment]
A few minor comments, but looks good otherwise. I will however let @caesarxuchao to add the LGTM label.

---

Reviewed 2 of 14 files at r1.
Review status: 2 of 22 files reviewed at latest revision, 11 unresolved discussions.

---

*[federation/apis/swagger-spec/apis.json, line 6 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31759#-KQTzfhL1wezClHBB_qO:-KQTzfhL1wezClHBB_qP:bfugzq0) ([raw file](https://github.com/kubernetes/kubernetes/blob/a954ff18e6e35414ccf5f82492b7c1349e4c522f/federation/apis/swagger-spec/apis.json#L6)):*
> ```JSON
>   "basePath": "https://10.10.10.10:6443",
>   "resourcePath": "/apis",
>   "info": {
> ```

Same as @caesarxuchao's previous comment.

---

*[federation/apis/swagger-spec/extensions.json, line 6 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31759#-KQTzkm03tiaVm6EdFcN:-KQTzkm03tiaVm6EdFcO:b2nreh6) ([raw file](https://github.com/kubernetes/kubernetes/blob/a954ff18e6e35414ccf5f82492b7c1349e4c522f/federation/apis/swagger-spec/extensions.json#L6)):*
> ```JSON
>   "basePath": "https://10.10.10.10:6443",
>   "resourcePath": "/apis/extensions",
>   "info": {
> ```

Same as above.

---

*[federation/apis/swagger-spec/extensions_v1beta1.json, line 6 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31759#-KQTzmkVuXXkRkpZikOK:-KQTzmkVuXXkRkpZikOL:b2nreh6) ([raw file](https://github.com/kubernetes/kubernetes/blob/a954ff18e6e35414ccf5f82492b7c1349e4c522f/federation/apis/swagger-spec/extensions_v1beta1.json#L6)):*
> ```JSON
>   "basePath": "https://10.10.10.10:6443",
>   "resourcePath": "/apis/extensions/v1beta1",
>   "info": {
> ```

Same as above.

---

*[federation/apis/swagger-spec/extensions_v1beta1.json, line 2858 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31759#-KQU-hN-iyndGru640ZF:-KQU-hN0nI-hXT1FHycb:b-d53bc3) ([raw file](https://github.com/kubernetes/kubernetes/blob/a954ff18e6e35414ccf5f82492b7c1349e4c522f/federation/apis/swagger-spec/extensions_v1beta1.json#L2858)):*
> ```JSON
>       "description": "VsphereVolume represents a vSphere volume attached and mounted on kubelets host machine"
>      },
>      "quobyte": {
> ```

A little curious about why this is here.

---

*[federation/apis/swagger-spec/federation.json, line 6 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31759#-KQU-oeU8Kru4TZ1xLUP:-KQU-oeU8Kru4TZ1xLUQ:b-dgr6fd) ([raw file](https://github.com/kubernetes/kubernetes/blob/a954ff18e6e35414ccf5f82492b7c1349e4c522f/federation/apis/swagger-spec/federation.json#L6)):*
> ```JSON
>   "basePath": "https://10.10.10.10:6443",
>   "resourcePath": "/apis/federation",
>   "info": {
> ```

Same comment as above and in all the many following occurrences.

---

*[hack/update-federation-generated-swagger-docs.sh, line 3 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31759#-KQU0Fy4b15XqLNqeUlK:-KQU0Fy4b15XqLNqeUlL:b-8k88q1) ([raw file](https://github.com/kubernetes/kubernetes/blob/a954ff18e6e35414ccf5f82492b7c1349e4c522f/hack/update-federation-generated-swagger-docs.sh#L3)):*
> ```Shell
> #!/bin/bash
> 
> # Copyright 2015 The Kubernetes Authors.
> ```

2016?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31759#-:-KQU0__JMIpRlDpRDfMq:b-t3ulyx)*

**Comment by madhusudancs** [comment]
> https://htmlpreview.github.io/?https://raw.githubusercontent.com/nikhiljindal/kubernetes/fedSwaggerDoc/federation/docs/api-reference/federation/v1beta1/operations.html

> https://htmlpreview.github.io/?https://raw.githubusercontent.com/nikhiljindal/kubernetes/fedSwaggerDoc/federation/docs/api-reference/federation/v1beta1/definitions.html

I see a stray `*versioned.Event` here. What is that supposed to be?

> https://htmlpreview.github.io/?https://raw.githubusercontent.com/nikhiljindal/kubernetes/fedSwaggerDoc/federation/docs/api-reference/extensions/v1beta1/operations.html

Articles preceding "Ingress" are all incorrect. For example, `create a Ingress` should be `create an Ingress`. Is there a way to fix that?

Do we allow reading the scale subresource in federation? If not we should remove it from this doc here.

Also, the headings for Scale subresources should be fixed. This is also a problem in Kubernetes reference docs, not specific to Federation. For example, `read scale of the specified Scale` should be `read scale of the specified ReplicaSet`. I am still not sure if that is correct. I don't know what it means to read "the scale" of a ReplicaSet, but I also don't know what's the correct formation.

Watch documentation ordering is pretty weird. (Also applies to Kubernetes). Any why we list watch in the end and everything else with the corresponding resources?

> https://htmlpreview.github.io/?https://raw.githubusercontent.com/nikhiljindal/kubernetes/fedSwaggerDoc/federation/docs/api-reference/extensions/v1beta1/definitions.html

> https://htmlpreview.github.io/?https://raw.githubusercontent.com/nikhiljindal/kubernetes/fedSwaggerDoc/federation/docs/api-reference/v1/operations.html


> https://htmlpreview.github.io/?https://raw.githubusercontent.com/nikhiljindal/kubernetes/fedSwaggerDoc/federation/docs/api-reference/v1/definitions.html

**Comment by nikhiljindal** [comment]
Review status: 2 of 22 files reviewed at latest revision, 11 unresolved discussions.

---

*[federation/apis/swagger-spec/extensions_v1beta1.json, line 2858 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31759#-KQU-hN-iyndGru640ZF:-KQWj2TI7mUBO7AtArm1:bx9gsbq) ([raw file](https://github.com/kubernetes/kubernetes/blob/a954ff18e6e35414ccf5f82492b7c1349e4c522f/federation/apis/swagger-spec/extensions_v1beta1.json#L2858)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary>
> A little curious about why this is here.
</details>
This must have been recently added to the PodTemplateSpec. Its here because ReplicaSetSpec includes PodTemplateSpec.

---

*[hack/update-api-reference-docs.sh, line 38 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31759#-KQTjy4AFu4MtJbwO5cL:-KQWi9OJ1fysUeg2T6x_:bb6un1j) ([raw file](https://github.com/kubernetes/kubernetes/blob/cda78f8338cd1445d73dd74c862690fbd779f80f/hack/update-api-reference-docs.sh#L38)):*
<details><summary><i>Previously, caesarxuchao (Chao Xu) wrote…</i></summary>
> This looks to be copied from the original code, but I think using kube::util::group-version-to-pkg-path will make it cleaner.
</details>
Done

---

*[hack/update-federation-api-reference-docs.sh, line 40 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31759#-KQTk7hK5JlcdC9ESAOj:-KQWiY7gmgVC7nhl7DE7:bb6un1j) ([raw file](https://github.com/kubernetes/kubernetes/blob/cda78f8338cd1445d73dd74c862690fbd779f80f/hack/update-federation-api-reference-docs.sh#L40)):*
<details><summary><i>Previously, caesarxuchao (Chao Xu) wrote…</i></summary>
> Same here, using kube::util::group-version-to-pkg-path seems better.
</details>
Done

---

*[hack/update-federation-generated-swagger-docs.sh, line 3 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31759#-KQU0Fy4b15XqLNqeUlK:-KQWkLcu9Oh0CO9PtSSf:b-886gsx) ([raw file](https://github.com/kubernetes/kubernetes/blob/a954ff18e6e35414ccf5f82492b7c1349e4c522f/hack/update-federation-generated-swagger-docs.sh#L3)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary>
> 2016?
</details>
Fixed

---

*[hack/lib/swagger.sh, line 79 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31759#-KQTke3HRTAmIwQpMBtD:-KQWic6Sm3jYuA1nR5Kj:bteqx5a) ([raw file](https://github.com/kubernetes/kubernetes/blob/cda78f8338cd1445d73dd74c862690fbd779f80f/hack/lib/swagger.sh#L79)):*
<details><summary><i>Previously, caesarxuchao (Chao Xu) wrote…</i></summary>
> Are these two lines redundant?
</details>
No. Without these GROUP_VERSIONS is a single string "v1 extensions/v1beta1". This command turns it into an array.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31759)*

File: hack/update-federation-generated-swagger-docs.sh (hunk 2 above)
Comment (nikhiljindal) [inline hack/update-federation-generated-swagger-docs.sh]: Done

File: federation/apis/swagger-spec/api.json (hunk 1 above)
Comment (nikhiljindal) [inline federation/apis/swagger-spec/api.json]: These got added as a result of https://github.com/kubernetes/kubernetes/pull/30914.

These fields are part of the spec, go-restful might have recently added support for it.

**Comment by nikhiljindal** [comment]
> > https://htmlpreview.github.io/?https://raw.githubusercontent.com/nikhiljindal/kubernetes/fedSwaggerDoc/federation/docs/api-reference/federation/v1beta1/operations.html
> > 
> > https://htmlpreview.github.io/?https://raw.githubusercontent.com/nikhiljindal/kubernetes/fedSwaggerDoc/federation/docs/api-reference/federation/v1beta1/definitions.html
> > I see a stray *versioned.Event here. What is that supposed to be?
> > 
> > https://htmlpreview.github.io/?https://raw.githubusercontent.com/nikhiljindal/kubernetes/fedSwaggerDoc/federation/docs/api-reference/extensions/v1beta1/operations.html
> 

> Articles preceding "Ingress" are all incorrect. For example, create a Ingress should be create an Ingress. Is there a way to fix that?

Yes it is possible to fix it. That doc is generated here: https://github.com/kubernetes/kubernetes/blob/106373123455afd5d6e71da809b85ad8f6455d9c/pkg/apiserver/api_installer.go#L599

> 
> Do we allow reading the scale subresource in federation? If not we should remove it from this doc here.

Yes, we support replicasets/scale: https://github.com/kubernetes/kubernetes/blob/106373123455afd5d6e71da809b85ad8f6455d9c/federation/cmd/federation-apiserver/app/extensions.go#L38

> 
> Also, the headings for Scale subresources should be fixed. This is also a problem in Kubernetes reference docs, not specific to Federation. For example, read scale of the specified Scale should be read scale of the specified ReplicaSet. I am still not sure if that is correct. I don't know what it means to read "the scale" of a ReplicaSet, but I also don't know what's the correct formation.
> 
> Watch documentation ordering is pretty weird. (Also applies to Kubernetes). Any why we list watch in the end and everything else with the corresponding resources?

**Comment by nikhiljindal** [comment]
Pushed a new commit as per comments, PTAL.

**Comment by madhusudancs** [comment]
Review status: 2 of 22 files reviewed at latest revision, 11 unresolved discussions.

---

*[federation/apis/swagger-spec/extensions_v1beta1.json, line 2858 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31759#-KQU-hN-iyndGru640ZF:-KQWozjPLi5liqY66i-4:b7tzviv) ([raw file](https://github.com/kubernetes/kubernetes/blob/a954ff18e6e35414ccf5f82492b7c1349e4c522f/federation/apis/swagger-spec/extensions_v1beta1.json#L2858)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary>
> This must have been recently added to the PodTemplateSpec. Its here because ReplicaSetSpec includes PodTemplateSpec.
</details>
Makes sense. 

Also, sorry I did not see this file's path correctly. Reviewable kind of makes it hard. I can't see the full path of the file I am reviewing until I hover on it. So I did not see this was in the federation directory. I was wondering why such a change was not already applied before when someone made the API change in Kubernetes. Now that I see `federation/...` in the path, it makes sense.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31759)*

**Comment by madhusudancs** [comment]
> > Articles preceding "Ingress" are all incorrect. For example, create a Ingress should be create an Ingress. Is there a way to fix that?

> Yes it is possible to fix it. That doc is generated here: https://github.com/kubernetes/kubernetes/blob/106373123455afd5d6e71da809b85ad8f6455d9c/pkg/apiserver/api_installer.go#L599

Fixing it there would change it for all the resources unless we actually semi-smart article detector right? My question was more on the lines of fixing it only for Ingress.

> > Do we allow reading the scale subresource in federation? If not we should remove it from this doc here.

> Yes, we support replicasets/scale: https://github.com/kubernetes/kubernetes/blob/106373123455afd5d6e71da809b85ad8f6455d9c/federation/cmd/federation-apiserver/app/extensions.go#L38

Cool! Thanks for confirming.

**Comment by nikhiljindal** [comment]
> > > Articles preceding "Ingress" are all incorrect. For example, create a Ingress should be create an Ingress. Is there a way to fix that?
> > 
> > Yes it is possible to fix it. That doc is generated here: https://github.com/kubernetes/kubernetes/blob/106373123455afd5d6e71da809b85ad8f6455d9c/pkg/apiserver/api_installer.go#L599
> > 
> 
> Fixing it there would change it for all the resources unless we actually semi-smart article detector right? My question was more on the lines of fixing it only for Ingress.

Yes, need to use an article detector. Other option will be to add an if there hardcoding ingress :)
But thats the place to set it appropriately.

**Comment by caesarxuchao** [comment]
Thank you both! LGTM.

**Comment by nikhiljindal** [comment]
Squashed the last commit, thanks!

**Comment by nikhiljindal** [comment]
Sent https://github.com/kubernetes/kubernetes.github.io/pull/1143 to import these docs on k8s.io

**Pull request merged**